File: 3. Behavioral Design Patterns/Visitor/Java/VisitorTemplate.java
// Visitor Pattern – Java Template

interface Visitor {
    void visit(ConcreteElementA element);
    void visit(ConcreteElementB element);
}

interface Element {
    void accept(Visitor visitor);
}

class ConcreteElementA implements Element {
    public void accept(Visitor visitor) { visitor.visit(this); }
    public void operationA() { /* specific to A */ }
}

class ConcreteElementB implements Element {
    public void accept(Visitor visitor) { visitor.visit(this); }
    public void operationB() { /* specific to B */ }
}

class ConcreteVisitor1 implements Visitor {
    public void visit(ConcreteElementA element) { element.operationA(); /* visitor 1 logic */ }
    public void visit(ConcreteElementB element) { element.operationB(); /* visitor 1 logic */ }
}

class ConcreteVisitor2 implements Visitor {
    public void visit(ConcreteElementA element) { element.operationA(); /* visitor 2 logic */ }
    public void visit(ConcreteElementB element) { element.operationB(); /* visitor 2 logic */ }
}

public class VisitorTemplate {
    public static void main(String[] args) {
        Element[] elements = { new ConcreteElementA(), new ConcreteElementB() };
        Visitor visitor = new ConcreteVisitor1();
        for (Element element : elements) element.accept(visitor);
    }
}

File: 3. Behavioral Design Patterns/Visitor/C#/VisitorTemplate.cs
// Visitor Pattern – C# Template

namespace VisitorTemplate
{
    interface IVisitor
    {
        void Visit(ConcreteElementA element);
        void Visit(ConcreteElementB element);
    }

    interface IElement
    {
        void Accept(IVisitor visitor);
    }

    class ConcreteElementA : IElement
    {
        public void Accept(IVisitor visitor) { visitor.Visit(this); }
        public void OperationA() { /* specific to A */ }
    }

    class ConcreteElementB : IElement
    {
        public void Accept(IVisitor visitor) { visitor.Visit(this); }
        public void OperationB() { /* specific to B */ }
    }

    class ConcreteVisitor1 : IVisitor
    {
        public void Visit(ConcreteElementA element) { element.OperationA(); /* visitor 1 logic */ }
        public void Visit(ConcreteElementB element) { element.OperationB(); /* visitor 1 logic */ }
    }

    class ConcreteVisitor2 : IVisitor
    {
        public void Visit(ConcreteElementA element) { element.OperationA(); /* visitor 2 logic */ }
        public void Visit(ConcreteElementB element) { element.OperationB(); /* visitor 2 logic */ }
    }

    class Program
    {
        static void Main()
        {
            IElement[] elements = { new ConcreteElementA(), new ConcreteElementB() };
            var visitor = new ConcreteVisitor1();
            foreach (var element in elements) element.Accept(visitor);
        }
    }
}
